packages/utilities/src/styled.tsx

```diff
@@ -46,8 +48,16 @@ export function styled<
   Component: React.ComponentClass<TComponentProps> | React.StatelessComponent<TComponentProps>,
   baseStyles: IStyleFunctionOrObject<TStyleProps, TStyleSet>,
   getProps?: (props: TComponentProps) => Partial<TComponentProps>,
-  customizable?: ICustomizableProps
+  customizable?: ICustomizableProps,
+  useMemoize?: boolean
 ): (props: TComponentProps) => JSX.Element {
+  let getStyleFunction = (...allStyles: (IStyleFunctionOrObject<TStyleProps, TStyleSet> | undefined)[]) => {
+    return (styleProps: TStyleProps) => _resolve(styleProps, ...allStyles);
+  };
+  if (useMemoize) {
+    getStyleFunction = memoizeFunction(getStyleFunction);
```

Comment:
So the way memoized functions work is that it looks at individual arguments to build up the tree of memoized responses.

In this case, we're memoizing a function which takes props as an argument. That means that the styles will be recalculated whenever the props object itself mutates.

It is extremely likely that the props for the component will be recomputed each time it is rendered. So I'm not sure this will actually work as you expect, without changing memoization to work on a deep object compare level.

e.g.

```ts
_counter = 0;

const foo = memoizeFunction((obj) => ++counter);
expect(foo({ a: 1})).toEqual(1);
expect(foo({ a: 1})).toEqual(1); // fails, because the object is a new instance.
```

Reply:
I had considered changing memoizeFunction to do a deep object comparison, or even a shallow... but then I worry if we're costing more time than saving.

Reply:
I know that if we do it this way the trick will not work:
```js
render() {
  return <Component styles={{ root: { height: this.props.height } }} />
}
```
But if we use some library like `reselect`, then `memoizeFunction` will work as expected:
```js
_componentStyleSelector = createSelector(
  (props) => props.height,
  (height) => ({root: { height }})
);

render() {
  return <Component styles={this._componentStyleSelector(this.props)} />;
}
```
What I want to do here is to enable this way of optimization.

Obviously if `memoizeFunction` could do some kind of deep comparing will help, but in my opinion that's not how react and redux handle data changes.

Reply:
ping @dzearing

Reply:
Sorry for the delay; i'm having a bit of a difficult time following this through to identify how it works and what might break. I need to enlist and debug through.

Reply:
Ok I pulled this down and see a problem.

memoizeFunction will look at the arguments passed in and build a graph where each node is an argument value, which has an edge to the next value, where the final value has a cached result.

In your test case, I debugged.

in styled, I see this

```tsx
 const styles = getStyleFunction(baseStyles, customizedStyles, componentProps.styles);
```

baseStyles (the function) gets passed in. In your memoized test, you cover rendering the same component twice and you validate that the styles don't get recalc'd. This is because you're passing the same function in the second time.

However the function will be evaluated, given the props. If anything in the props mutate, the function is not recalculated. So if you modify your test to, for example, not pass in the `cool` prop, the styles aren't recalculated.

```ts
   component.update(<TestCached cool />);
    expect(component.toJSON()).toMatchSnapshot();
    expect(_lastProps!.styles).toBe(originStyles);
    originStyles = _lastProps!.styles;

    component.update(<TestCached />);
    expect(component.toJSON()).toMatchSnapshot();
// ====== THIS NEXT LINE WILL FAIL ==================
    expect(_lastProps!.styles).not.toBe(originStyles);
// ============================================
    originStyles = _lastProps!.styles;
```

So the only way this can work is for things which only receive static objects, rather than style functions.

In your style function, which takes props, you're free to use a memoized function within there, using normalized params which can build the memoization graph correctly:

```ts
function getStyles = memoizedFunction((theme, cool) => ({
  background: cool ? theme.palette.themePrimary : theme.palette.themeSecondary
}));

styled(
  BaseComponent,
  (props) => getStyles(props.theme, props.cool) // normalized values can be memoized correctly
)
```

Reply:
I would recommend we avoid baking this logic into styled. It requires the component author to choose which props are immutable and should be pushed into memoization evaluation.

What do you think?

Reply:
> However the function will be evaluated, given the props. If anything in the props mutate, the function is not recalculated. So if you modify your test to, for example, not pass in the cool prop, the styles aren't recalculated.

That's expected behavior I think, for that the component received `props.styles` is a function, and the component should resolve the final style sheet itself with `getClassNames()`. I've updated the test to cover this scenario and, as you see, it works perfectly (`props.styles` doesn't change and rendered styles changed if `props.cool` changes).

> you're free to use a memoized function within there, using normalized params which can build the memoization graph correctly

This doesn't work because `styled()` keeps generating different `props.styles` (function) even if the same styles is received.

Reply:
One thing hits me that the `settings` or `theme` extracted from global context may not work as I expected. I'll check it tomorrow.